keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/NotificationEventListener.kt

```diff
@@ -26,64 +35,180 @@ class NotificationEventListener(
+          time = clock.instant(),
+          application = config.name,
+          comment = veto.comment
+        ),
+        Type.ARTIFACT_MARK_AS_BAD,
+        veto.targetEnvironment
+      )
+    }
+  }
+
+  @EventListener(ApplicationActuationPaused::class)
+  fun onApplicationActuationPaused(notification: ApplicationActuationPaused) {
+    with(notification) {
+      val config = repository.getDeliveryConfigForApplication(application)
+
+      sendSlackMessage(config,
+        SlackPausedNotification(
+        user = triggeredBy,
+        time = clock.instant(),
+        application = application
+      ),
+        Type.APPLICATION_PAUSED)
+    }
+
+  }
+
+  @EventListener(ApplicationActuationResumed::class)
+  fun onApplicationActuationResumed(notification: ApplicationActuationResumed) {
+    with(notification) {
+      val config = repository.getDeliveryConfigForApplication(application)
+
+      sendSlackMessage(config,
+        SlackResumedNotification(
+        user = triggeredBy,
+        time = clock.instant(),
+        application = application
+      ),
+        Type.APPLICATION_RESUMED)
+    }
+  }
+
+  @EventListener(LifecycleEvent::class)
+  fun onLifecycleEvent(notification: LifecycleEvent) {
+    with(notification) {
+      val config = repository.getDeliveryConfig(artifactRef.split(":")[0])
+      val deliveryArtifact = config.artifacts.find {
+        it.reference == artifactRef.split(":")[1]
+      } ?: return
+
+      val artifact = repository.getArtifactVersion(deliveryArtifact, artifactVersion, null)
+      if (artifact == null) {
+        log.debug("delivery artifact is null for application ${config.application}. Can't send $type notification")
+        return
+      }
+
+      if (status == LifecycleEventStatus.FAILED) {
+        sendSlackMessage(config,
+          SlackLifecycleNotification(
+          time = clock.instant(),
+          artifact = artifact,
+          eventType = type,
+          application = config.application
+        ),
+          Type.LIFECYCLE_EVENT,
+          artifact = deliveryArtifact)
       }
     }
   }
+
+  private inline fun <reified T : SlackNotificationEvent> sendSlackMessage(config: DeliveryConfig, message: T, type: Type,
+                                                                           targetEnvironment: String? = null,
+                                                                           artifact: DeliveryArtifact? = null) {
+    val handler: SlackNotificationHandler<T>? = handlers.supporting(type)
+
+    if (handler == null) {
+      log.debug("no handler was found for notification type ${T::class.java}. Can't send slack notification.")
+      return
+    }
+
+    //if targetEnvironment is not null, use only its notifications. Else, use all notifications configured for all environments.
+    val environments: Set<Environment> = config.environments.filter {
+      targetEnvironment == null || it.name == targetEnvironment
+    }.also { it ->
```

Comment:
@luispollo @emjburns I added these filters as part of your comments, but I think I have a small bug here.. `artifact.isUsedIn(it)` is returning true even if the `resource` is empty 🤷 , and the test below is failing. any ideas?

Reply:
I'm looking, but in the meantime, you could simplify this by removing the `.also` block and just chaining the other `.filter`.

Reply:
@luispollo that was it!!! fixed.